R/pass_fail_condi.R

```diff
@@ -0,0 +1,65 @@
+# Fail if condition matches
+#' @export
+fail_if <- function(x, message) {
+  condi(x, message, correct = FALSE)
+}
+
+#' Pass if condition matches
+#' @export
+pass_if <- function(x, message) {
+  condi(x, message, correct = TRUE)
+}
+
+#' Condition object
+#' Captures what the user passes into \code{pass_if} or \code{fail_if},
+#' figures out what type of object was passed into \code{x},
+#' and returns an object that will be passed into \code{evaluate_condi}
+condi <- function(x, message, correct) {
+  type <-
+    if (rlang::is_formula(x)) {
+      "formula"
+    } else if (rlang::is_function(x)) {
+      "function"
+    } else {
+      "value"
+    }
+
+  ret <- list(
+    x = x,
+    message = message,
+    correct = correct,
+    type = type
+  )
+  class(ret) <- "grader_condition"
+  ret
+}
+
+#' Evaluates a condition
+#' @returns Boolean of the condition
+#' @export
+evaluate_condi <- function(condi, grader_args, learnr_args) {
+  checkmate::assert_class(condi, "grader_condition")
+  switch(condi$type,
```

Comment:
wrap this in a `try(..., silent = TRUE)`. If the result inherits a `try-error`, then return a `graded(FALSE, NULL)`.  

(Comparisons could cause an error. Function evaluation could cause an error. Good to be safe.)

Reply:
inside the `try`, If the value is TRUE, return  `graded(condi$correct, condi$message)`.  Otherwise return NULL

Reply:
If the value is not a boolean or `graded` output, throw an error.

Reply:
The `NULL` is causing a few problems... Correct me if I'm wrong and might be thinking about this the wrong way...

Essentially `res` acts as a "did I find a match?" and if there was an error in running the `x` we return `graded(correct = FALSE, message = NULL)`.

So, we shouldn't need to check if `res` is `NULL` since that doesn't seem will ever happen (it should only be a T/F value or a `graded` object).

Next, since `res` is a logical:
if it's true (we found a match), then we return `return(graded(correct = condi$correct, message = condi$message))`
which is fine... but if we do not find a match we should return something else (maybe `NULL` makes sense only in this location?).

I say this because when we leave `evaluate_condi` and go back to `check_result`, `evaluated_condi` needs to account for a matched `fail_if` call. Right now if `fail_if` does match, we never do anything about it

```r
if (!is.null(evaluate_condi) && evaluated_condi$correct) {
          found_match <- TRUE
          final_result <- evaluated_condi
          break
        }
```

Reply:
discussing no longer relevant. we removed the tryCatch and if the student's solution is wrong, we return `NULL` this lets us check for a found match without having to also check for a correct solution